### PR TITLE
docs: add design document for foreign data wrapper

### PR DESCRIPTION
**Pull request by iguoyr**
### What problem does this PR solve?

Issue Number: https://github.com/pingcap/tidb/issues/18677 

Problem Summary: make TiDB able to manage external data

### What is changed and how it works?

Proposal: [Support Foreign Data Wrapper](https://github.com/iguoyr/tidb/blob/fdw-rfc/docs/design/2021-02-01-foreign-data-wrapper.md) 

What's Changed: design document

### Check List 

Tests 

- No code

### Release note 

- No release note

### CODE REVIEW

**Review by zz-jason — CHANGES_REQUESTED** [review]
please write the design doc in English

**Comment by TisonKun** [comment]
Thanks for your contribution @iguoyr @wph95 @lfkdsk ! Generally we would like to accept RFC in English only and here is an English [proposal](https://docs.google.com/document/d/1YYjIfGznIgUXA06fuh3KL1YFx2R1hVyu3duoR3GM42o) on foreign-data wrapper written recently.

Your works are highly appreciated and I'd like to share my concerns on foreign-data wrapper topic. Let's cooperate on both  google doc linked above for the draft of this proposal and materialize it into this pr and thus docs/design later we reach a consensus.

Below are my concerns and information from various channel so far. Let's discuss on the questions.

**Comment by TisonKun** [comment]
1. Which plugin framework we'd like to choose?

As mentioned in [current design](https://github.com/pingcap/tidb/blob/f404c4121935f64e2a4e51273a8a1dd2b8861357/docs/design/2021-02-01-foreign-data-wrapper.md), it is proposed that we use the existing TiDB plugin framework to support foreign-data wrapper requirement. But also as mentioned in the doc, due to go plugin limitation, the plugin must be recompiled and reloaded when the source code tree changed.

During offline discussion @sunxiaoguang proposed a workaround that we implement two meta plugins in TiDB plugin framework, one adapts cgo plugins and the other adapts gRPC plugins.

**Comment by TisonKun** [comment]
2. Which SQL grammar we'd like to support create foreign table?

Foreign table, or external table, is one of the key concepts when supports foreign-data wrapper. There are quite a lot implementation to refer. I do a bit hesitate on which grammar is the most smooth for both TiDB integrates and user implements.

**Comment by TisonKun** [comment]
2.1 The SQL/MED grammar is as below, which is examined over years and supports very flexible callback points. However, the "FOREIGN DATA WRAPPER" -> "SERVER" -> "USER MAPPING" abstraction is a bit obsolete from modern perspective. Also the original FDW implementation in PostgreSQL focuses a lot on privilege things, which might not be the most significant things we bring foreign-data wrapper into TiDB.

```
-- load mysql_fdw first time after install
CREATE FOREIGN DATA WRAPPER mysql_fdw;

-- create server object
CREATE SERVER mysql_server
	FOREIGN DATA WRAPPER mysql_fdw
	OPTIONS (host '127.0.0.1', port '3306');

-- create user mapping
CREATE USER MAPPING FOR root
	SERVER mysql_server
	OPTIONS (username 'foo', password 'bar');

-- create foreign table
CREATE FOREIGN TABLE warehouse
	(
		warehouse_id int,
		warehouse_name text,
		warehouse_created timestamp
	)
	SERVER mysql_server
	OPTIONS (dbname 'db', table_name 'warehouse');

-- insert new rows in table
INSERT INTO warehouse values (1, 'UPS', current_date);
INSERT INTO warehouse values (2, 'TV', current_date);
INSERT INTO warehouse values (3, 'Table', current_date);

-- select from table
SELECT * FROM warehouse ORDER BY 1;

warehouse_id | warehouse_name | warehouse_created
-------------+----------------+-------------------
           1 | UPS            | 10-JUL-20 00:00:00
           2 | TV             | 10-JUL-20 00:00:00
           3 | Table          | 10-JUL-20 00:00:00

-- delete row from table
DELETE FROM warehouse where warehouse_id = 3;

-- update a row of table
UPDATE warehouse set warehouse_name = 'UPS_NEW' where warehouse_id = 1;

-- explain a table with analyze option
EXPLAIN ANALYZE SELECT warehouse_id, warehouse_name FROM warehouse WHERE warehouse_name LIKE 'TV' limit 1;
```

**Comment by TisonKun** [comment]
2.2 Extended ENGINE grammar is as below, which is chosen to integrate foreign table [by ClickHouse](https://clickhouse.tech/docs/en/engines/table-engines/integrations/mysql/).

```
CREATE TABLE [IF NOT EXISTS] [db.]table_name [ON CLUSTER cluster]
(
    name1 [type1] [DEFAULT|MATERIALIZED|ALIAS expr1],
    name2 [type2] [DEFAULT|MATERIALIZED|ALIAS expr2],
    ...
) ENGINE = Kafka()
SETTINGS
    kafka_broker_list = 'host:port',
    kafka_topic_list = 'topic1,topic2,...',
    kafka_group_name = 'group_name',
    kafka_format = 'data_format'[,]
    [kafka_row_delimiter = 'delimiter_symbol',]
    [kafka_schema = '',]
    [kafka_num_consumers = N,]
    [kafka_max_block_size = 0,]
    [kafka_skip_broken_messages = N,]
    [kafka_commit_every_batch = 0,]
    [kafka_thread_per_consumer = 0]
```

**Comment by TisonKun** [comment]
2.3 Introduce a WITH statement populated with all settings required for the foreign table. This is [how Flink implements it](https://ci.apache.org/projects/flink/flink-docs-release-1.12/dev/table/sql/create.html).

```
CREATE TABLE MyUserTable (
  column_name1 INT,
  column_name2 STRING,
  ...
  part_name1 INT,
  part_name2 STRING
) PARTITIONED BY (part_name1, part_name2) WITH (
  'connector' = 'filesystem',           -- required: specify the connector
  'path' = 'file:///path/to/whatever',  -- required: path to a directory
  'format' = '...',                     -- required: file system connector requires to specify a format,
                                        -- Please refer to Table Formats
                                        -- section for more details
  'partition.default-name' = '...',     -- optional: default partition name in case the dynamic partition
                                        -- column value is null/empty string

  -- optional: the option to enable shuffle data by dynamic partition fields in sink phase, this can greatly
  -- reduce the number of file for filesystem sink but may lead data skew, the default value is false.
  'sink.shuffle-by-partition.enable' = '...',
  ...
)
```

**Comment by TisonKun** [comment]
I'd prefer option 2.2 since it is extended existing MySQL grammar ENGINE and keep it simple.

Other alternatives is omitted here because they grow their own complex grammar to support specific scenarios such as Spark / Calcite /  Presto / Materialize. However, their use facing interface is valuable to refer. I will start this discussion below.

**Comment by TisonKun** [comment]
3. FDW API

This is the most important part for both TiDB developers invoking FDW logics and official unofficial FDW implementation following.

I'd like to share PostgreSQL FDW callbacks definition and how TiKV / TiFlash coprocessor works.

Below is PostgreSQL FDW callbacks definition. See more information in [this page](https://www.postgresql.org/docs/current/fdw-callbacks.html)

```C
// scan
void GetForeignRelSize;
void GetForeignPaths;
ForeignScan * GetForeignPlan;
void BeginForeignScan;
TupleTableSlot * IterateForeignScan;
void ReScanForeignScan;
void EndForeignScan;

// join
void GetForeignJoinPaths;

// post-scan / join processing
void GetForeignUpperPaths;

// update
void AddForeignUpdateTargets;
List * PlanForeignModify;
void BeginForeignModify;
TupleTableSlot * ExecForeignInsert;
TupleTableSlot * ExecForeignUpdate;
TupleTableSlot * ExecForeignDelete;
void EndForeignModify;
void BeginForeignInsert;
void EndForeignInsert;
int IsForeignRelUpdatable;
bool PlanDirectModify;
void BeginDirectModify;
TupleTableSlot * IterateDirectModify;
void EndDirectModify;

// row locking
RowMarkType GetForeignRowMarkType;
void RefetchForeignRow;
bool RecheckForeignScan;

// explain
void ExplainForeignScan;
void ExplainForeignModify;
void ExplainDirectModify

// analyze
bool AnalyzeForeignTable;
int AcquireSampleRowsFunc;

// ...
```

You can see that they are grouped by different scenario such as table scan, join, (aggregation) function, and so on, and they are tightly coupled with PostgreSQL implementation.

Thus, it's hard to just mapping these functions into TiDB FDW design, and the same as [current design](https://github.com/pingcap/tidb/blob/f404c4121935f64e2a4e51273a8a1dd2b8861357/docs/design/2021-02-01-foreign-data-wrapper.md), because the actual TiDB planning phase works with a `TableReader` which keeps executing `Next` for pulling data from engine.

To integrate with TiDB planning phase as well as fit current fetching data model, I propose APIs as below.

```
bool acceptLogicalPlan(LogicalPlan);
void visitDAGRequest(DAGRequest);
```

and the FDW reacts if the `LogicalPlan` can be pushed down and when the final plan constructed, handle the DAGRequest pushed down to it.

File: docs/design/2021-02-01-foreign-data-wrapper.md [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+# 标准 FDW 方式创建 使用 csv engine 的table
+CREATE EXTENSION csv_fdw;
+CREATE SERVER test_logs FOREIGN DATA WRAPPER csv_fdw;
+CREATE FOREIGN TABLE people (city int, name char(255)) SERVER test_logs
+OPTIONS ( filename '<YOUR_CSV_FILE_PATH>');
+
+# 使用 MySQL 兼容方案创建 table
+CREATE TABLE people(city int, name char(255)) ENGINE = csv ENGINE_ATTRIBUTE = "path=<YOUR_CSV_FILE_PATH>"
+  
+# 支持 insert
```
Comment (ilovesoup) [inline docs/design/2021-02-01-foreign-data-wrapper.md:47]: 暂时我没有特别看到 API 写入支持的好处？这里担心的点是支持写入会大大增加复杂度。

File: docs/design/2021-02-01-foreign-data-wrapper.md [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+INSERT INTO people values(1, 'wph95');
+INSERT INTO people values(3, 'iguory');
+INSERT INTO people values(0, 'lfkdsk');
+
+# 支持 select
+SELECT * FROM people;
+SELECT * FROM people where city = 2;
+
+# 支持与 TiDB 普通 table: city 进行 join 
+SELECT city.cityName, people.name FROM city INNER JOIN people ON city.id=people.city
```
Comment (ilovesoup) [inline docs/design/2021-02-01-foreign-data-wrapper.md:57]: This won't be very special even if we just support table scan. The complex part might be if we support join within FD engine if it is a database.

File: docs/design/2021-02-01-foreign-data-wrapper.md [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+- `OnUpdate` 用于更新 foreign table 中的数据
+
+- `OnSelect` 用于下推 scan 算子 
+
+### DDL
+
+- `OnCreateServer` 用于创建 Server, `CREATE SERVER {server_name}` 时被调用
+- `OnDropServer` 用于删除 Server, `Drop SERVER {server_name}` 时被调用
+- `OnCreateTable` 用于创建 table，`CREATE TABLE {table_name}`时被调用
+- `OnDropTable` 用于删除 table，`DROP TABKE {table_name}`时被调用
```
Comment (ilovesoup) [inline docs/design/2021-02-01-foreign-data-wrapper.md:148]: Drop the link or drop physical table?

File: docs/design/2021-02-01-foreign-data-wrapper.md [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+## Compatibility and Limitations
+
+虽然本提案的计划只对 TiDB 主分支代码进行一些 Hook 埋点。作为一种对 Mysql SQL 的扩展，对已有功能未有任何修改与破坏，不存在兼容性问题。
+
+但因为以下原因，需要额外工作量：
+
+- 部分库和 Plugin 存在循环引用，需要针对性调整。
+- Mysql 的语法中并没有针对 `SQL/MED` 的设计，本提案遵循了 PostgreSQL 的语法与实现细节。同时将参考 MYSQL engine 的相关语法，针对性的进行 FDW 兼容。
```
Comment (ilovesoup) [inline docs/design/2021-02-01-foreign-data-wrapper.md:157]: If there is no SQL/MED in MySQL, we might consider not using PG design but consider more of newer federated query engine like presto / spark or even dremio.

**Review by zz-jason — COMMENTED** [review]
@TisonKun 

I prefer to only support read (query) operations on FDW instances in the first stage of FDW API v1.0.

1. Privilege is important, especially in the cloud infrastructure and data federation scenarios
2. The API should be able to share the capability of FDW instances with TiDB:
	- Tables can be accessed and their schema, stats (optional).
	- Supported SQL functions and operators. Calcite does by implementing adaptor rules for a specific adaptor (FDW instance)
3. The API should be able to be involved during query execution, in a volcano-like processing model, these execution options are open-next-close

**Comment by TisonKun** [comment]
@zz-jason thanks for your reply! My comment is below inline.

> I prefer to only support read (query) operations on FDW instances in the first stage of FDW API v1.0.

I agree. Support read only prevent us from complex transaction requirement. And federation benefits mainly on join multiple data source.

> 1. Privilege is important, especially in the cloud infrastructure and data federation scenarios

For privilege topic, we actually have two concepts on privilege: one guards who is able to read the external table's metadata, and the other guards if the user has the privilege to the metadata can access the external table physically.

For example, only the user create the external table can access the external table's metadata, and only if the metadata is correct, e.g. MySQL user & password, the user can access the external table physically with metadata.

If we follow PostgreSQL FDW totally, we meet the complex privilege solution on (1) FOREIGN DATA WRAPPER (2) SERVER (3) USER MAPPING (4) TABLE. And thus I'd prefer SQL syntax flattens metadata when creating external table.

> 2. The API should be able to share the capability of FDW instances with TiDB:
>    
>    * Tables can be accessed and their schema, stats (optional).
>    * Supported SQL functions and operators. Calcite does by implementing adaptor rules for a specific adaptor (FDW instance)
> 3. The API should be able to be involved during query execution, in a volcano-like processing model, these execution options are open-next-close

these topic is still under consideration, will comment later.

File: docs/design/2021-02-01-foreign-data-wrapper.md (hunk 1 above)
Comment (TisonKun) [reply docs/design/2021-02-01-foreign-data-wrapper.md:47]: +1 @iguoyr what do you think?

**Comment by marsishandsome** [comment]
>  And thus I'd prefer SQL syntax flattens metadata when creating external table.

Let's consider the following situation:
- mysql password is stored with the foreign table metadata
- a user is only granted with reading permission to the mysql foreign table
- the user run `show create table` command and get the mysql password
- the user now can connect the mysql instance with the password and may have writing permission

**Comment by tisonkun** [comment]
> > And thus I'd prefer SQL syntax flattens metadata when creating external table.
> 
> Let's consider the following situation:
> 
> * mysql password is stored with the foreign table metadata
> * a user is only granted with reading permission to the mysql foreign table
> * the user run `show create table` command and get the mysql password
> * the user now can connect the mysql instance with the password and may have writing permission
> 
> `(1) FOREIGN DATA WRAPPER (2) SERVER (3) USER MAPPING (4) TABLE` is a complex privilege solution, but secure.

@marsishandsome I agree the privilege part should be elaborated.

1. ClickHouse has [a privilege](https://clickhouse.tech/docs/en/sql-reference/statements/grant/#grant-show) dedicated for "SHOW CREATE TABLE" and I think it is different from SELECT privilege while on VIEW MySQL actually have Show views privilege.
2. Databrick SQL Analytics return options including username & password when querying SHOW CREATE TABLE, see also [this page](https://docs.databricks.com/sql/language-manual/sql-ref-syntax-aux-show-create-table.html).

I will think of the privilege's staff and concept separated between server & table, but fdw can be an engine extension with lookup and user mapping is a bit overkill(for privilege thought, inline into server is enough and we don't leak the server information if one only want to read data from the table).

File: docs/design/2021-02-01-foreign-data-wrapper.md [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+ OnCreateTable      func(sv *model.ServerInfo, tb *model.TableInfo) error
+ OnDropTable        func(tb *model.TableInfo) error
+}
+
+```
+
+### 算子
+
+- `OnInsert` 用于对 Foreign Table 写入数据
```
Comment (bb7133) [inline docs/design/2021-02-01-foreign-data-wrapper.md:133]: Do we expect ACID for the foreign table?

**Comment by tisonkun** [comment]
@zz-jason @marsishandsome here is the summary of my investigation of foreign table privileges problem.

1. For the access to data in foreign table, we keep current privilege model.
2. For the privacy leak risk introduce by "SHOW CREATE TABLE t" and "CREATE EXTERNAL TABLE t ( ... ) WITH ( ... )", we setup a gradual config strategy.
  a. we support config all options with "WITH" statements and pass them to fdw for create a connection(handle) to foreign table. ( refer - Spark / Flink )
  b. alternatively, we support CREATE SERVER and CREATE CREDENTIAL for create a foreign server or specific credential used to connect foreign server / table.

2.b turns into SQL syntax as below.

```sql
CREATE CREDENTIAL credential_name WITH ( 'key1' = 'value1, ...);
CREATE SERVER server_name WITH ( 'key1' = 'value1', ...);
CREATE SERVER server_name WITH ( 'credential' = 'credential_name', ...); -- optional
CREATE EXTERNAL TABLE table_name WITH ( 'key1' = 'value1', ...);
CREATE EXTERNAL TABLE table_name WITH ( 'server' = 'server_name;, ...); -- optional
CREATE EXTERNAL TABLE table_name WITH ( 'credential' = 'credential_name',  ...); -- optional
```

Meanwhile, we provide functions in fdw toolkits to get configs of server / credential by its name.

The considerations to this design are as below.

1. Community users do not always require strong privilege management. We can have a agile development path with just pack all configs.
2. Advanced features can be developed in decoupled, staged way.
3. For CREATE SERVER, I'm thinking of supporting syntax like IMPORT FOREIGN SCHEMA to batch create external tables.

x. we should think carefully about the name of concept especially "SERVER".

**Comment by marsishandsome** [comment]
> CREATE CREDENTIAL credential_name WITH ( 'key1' = 'value1, ...);

is there any database that uses `CREATE CREDENTIAL` syntax?

> x. we should think carefully about the name of concept especially "SERVER".

`SERVER` is a perfect abstraction to my opinion.  For example, we have a mysql server and 3 foreign tables.

```
CREATE SERVER mysql_server OPTIONS (host '127.0.0.1', port '3306');

CREATE FOREIGN TABLE t1 SERVER mysql_server ...;
CREATE FOREIGN TABLE t2 SERVER mysql_server ...;
CREATE FOREIGN TABLE t3 SERVER mysql_server ...;
```

If the myql ip changes from `127.0.0.1` to `127.0.0.2`,  
- we do not need to change the 3 foreign table 
- instead we only need to update the `SERVER` by running `alter SERVER mysql_server`

**Comment by tisonkun** [comment]
@marsishandsome thanks for your insights.

> > CREATE CREDENTIAL credential_name WITH ( 'key1' = 'value1, ...);
> 
> is there any database that uses `CREATE CREDENTIAL` syntax?
> 
> what's the differences between `CREATE CREDENTIAL` and `CREATE USER MAPPING` ?

[CREATE CREDENTIAL (Transact-SQL)](https://docs.microsoft.com/en-us/sql/t-sql/statements/create-credential-transact-sql?view=sql-server-ver15), you can think of the name and if we don't follow SQL/MED totally, the name is flexible. It is something likes USER MAPPING but shorter and having obvious semantic on privilege.

> > x. we should think carefully about the name of concept especially "SERVER".
> 
> `SERVER` is a perfect abstraction to my opinion. For example, we have a mysql server and 3 foreign tables.
> 
> ```
> CREATE SERVER mysql_server OPTIONS (host '127.0.0.1', port '3306');
> 
> CREATE FOREIGN TABLE t1 SERVER mysql_server ...;
> CREATE FOREIGN TABLE t2 SERVER mysql_server ...;
> CREATE FOREIGN TABLE t3 SERVER mysql_server ...;
> ```
> 
> If the myql ip changes from `127.0.0.1` to `127.0.0.2`,
> 
> * we do not need to change the 3 foreign tables
> * instead we only need to update the `SERVER` by running `alter SERVER mysql_server`

I agree the power brought by the "SERVER" abstraction. The problem is that SERVER is a bit ambiguous whether corresponds to a database or a dbms(catalog) that has multiple databases.

BTW, I'd prefer a consistent syntax `CREATE EXTERNAL TABLE t1 WITH ( 'server' = 'mysql_server' )` and provides a utils `get_server(server_name)` or `get_server_settings(server_name)` for fdw developers.

**Comment by tisonkun** [comment]
@wph95 @iguory I'd like to create another pr taking over the foreign data wrapper work if there is no activity in this week.

**Comment by tisonkun** [comment]
Closed as the pr author inactive for a while. Also I'm working on other topic so that a design doc based this one postpone.

You're welcome to reopen the PR or create a new one, and squash comments above to revive this topic.

Thanks for your contribution!

**Pull request closed without merging**